compiler/annotation/type_seeder.rs

```diff
@@ -672,6 +675,47 @@ impl UnaryConstraint for RoleName<Variable> {
     }
 }
 
+impl UnaryConstraint for Value<Variable> {
+    fn apply<Snapshot: ReadableSnapshot>(
+        &self,
+        seeder: &TypeGraphSeedingContext<'_, Snapshot>,
+        graph_vertices: &mut VertexAnnotations,
+    ) -> Result<(), TypeInferenceError> {
+        let pattern_value_type = match self.value_type() {
+            ir::pattern::ValueType::Builtin(value_type) => Ok(value_type.clone()),
+            ir::pattern::ValueType::Struct(struct_name) => {
+                let pattern_key = seeder.type_manager.get_struct_definition_key(seeder.snapshot, struct_name);
+                match pattern_key {
+                    Ok(Some(key)) => Ok(ValueType::Struct(key)),
+                    Ok(None) => Err(TypeInferenceError::ValueTypeNotFound { name: struct_name.clone().to_owned() }),
+                    Err(source) => Err(TypeInferenceError::ConceptRead { source }),
+                }
+            }
+        }?;
+
+        let mut annotations = BTreeSet::new();
+        seeder
+            .type_manager
+            .get_attribute_types(seeder.snapshot)
+            .map_err(|source| TypeInferenceError::ConceptRead { source })?
+            .into_iter()
+            .try_for_each(|attribute_type| {
+                let attribute_value_type_opt = attribute_type
+                    .get_value_type_without_source(seeder.snapshot, seeder.type_manager)
+                    .map_err(|source| TypeInferenceError::ConceptRead { source })?;
+                if let Some(attribute_value_type) = attribute_value_type_opt {
+                    if pattern_value_type == attribute_value_type {
+                        annotations.insert(TypeAnnotation::Attribute(attribute_type));
+                    }
+                }
+                Ok(())
+            })?;
```

Comment:
This can be written a little simpler with less nesting.
```suggestion
        let attribute_types = seeder
            .type_manager
            .get_attribute_types(seeder.snapshot)
            .map_err(|source| TypeInferenceError::ConceptRead { source })?;
        for attribute_type in attribute_types {
            let attribute_value_type = attribute_type
                .get_value_type_without_source(seeder.snapshot, seeder.type_manager)
                .map_err(|source| TypeInferenceError::ConceptRead { source })?;
            if attribute_value_type == Some(pattern_value_type) {
                annotations.insert(TypeAnnotation::Attribute(attribute_type));
            }
        }
```

Reply:
Oh wow, I didn't think about using `attribute_value_type == Some(pattern_value_type)`!! That's fire, thanks.